core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -114,36 +114,77 @@ namespace Experimental
 namespace Impl
 {
 
-// Custom callback for ArborX::BVH::query()
 template <typename Tag>
-struct NeighborDiscriminatorCallback;
+struct CollisionFilter;
+
+template <>
+struct CollisionFilter<FullNeighborTag>
+{
+    KOKKOS_FUNCTION bool static keep( int i, int j ) noexcept
+    {
+        return i != j; // discard self-collision
+    }
+};
 
 template <>
-struct NeighborDiscriminatorCallback<Cabana::FullNeighborTag>
+struct CollisionFilter<HalfNeighborTag>
+{
+    KOKKOS_FUNCTION static bool keep( int i, int j ) noexcept { return i > j; }
+};
+
+// Custom callback for ArborX::BVH::query()
+template <typename Tag>
+struct NeighborDiscriminatorCallback
 {
     using tag = ArborX::Details::InlineCallbackTag;
     template <typename Predicate, typename OutputFunctor>
-    KOKKOS_FUNCTION void operator()( Predicate const &predicate, int i,
+    KOKKOS_FUNCTION void operator()( Predicate const &predicate,
+                                     int primitive_index,
                                      OutputFunctor const &out ) const
     {
-        if ( getData( predicate ) != i ) // discard self-collision
+        int const predicate_index = getData( predicate );
+        if ( CollisionFilter<Tag>::keep( predicate_index, primitive_index ) )
         {
-            out( i );
+            out( primitive_index );
         }
     }
 };
 
-template <>
-struct NeighborDiscriminatorCallback<Cabana::HalfNeighborTag>
+template <typename Counts, typename Tag>
+struct NeighborDiscriminatorCallback2D_FirstPass
```

Comment:
By `FirstPass` and `SecondPass` do you mean these are called during the count and fill phases in ArborX? Is there something in the type tags that lets ArborX decide which functor to call when?

Reply:
I figured this out - see comments below

Reply:
I added comments above the specialization to make this more obvious.